Take adjustments into account in discount calculator

Noah-Silvera opened the pull request:
What is the goal of this PR?
---

TaxJar requires the line item totals to match the order total when posting transactions. 

We need to ensure any adjustments that are not tax are included in the line item discount.

Merge Checklist
---

- [x] Run the manual tests
- [x] Run the scenario in https://github.com/SuperGoodSoft/solidus_taxjar/issues/184 and ensure it is fixed.
- [x] Update the changelog

Noah-Silvera committed:
Take adjustments into account in discount calculator

TaxJar requires the line item totals to match the order total when
posting transactions. If you had canceled an item, and submitted a
refund for the cancelled item, re-posting that order to TaxJar would
fail, as the line_item would have the promo_total deducted but not the
adjustment made to the line item when it was cancelled.

We need to ensure any adjustments that are not tax are included in
the line item discount.

Co-authored-by: Nick Van Doorn <nick@super.gd>
Co-authored-by: Andrew Stewart <andrew@super.gd>
Co-authored-by: Chris Todorov <chris@super.gd>

forkata commented on lib/super_good/solidus_taxjar/discount_calculator.rb:
```diff
@@ -6,7 +6,7 @@ def initialize(line_item)
 
       def discount
-        -line_item.promo_total
+        -(line_item.promo_total + line_item.adjustments.select { |value| !value.tax? && value.eligible? }.sum(&:amount))
```
Me and @nvandoorn expanded on the tests added here and discovered that this will double count promotion adjustments since they are already included in the `promo_total`. We have an updated version of this change which completely removes the `promo_total` and just does the sum of the adjustments in Ruby which may be closer to what we want.

Noah-Silvera commented:
Updated to take into account all non-tax adjustment types in the discount calculation!

Noah-Silvera commented:
Just added a test for non-eligible adjustments as we were missing one!

Noah-Silvera committed:
Update CHANGELOG

nvandoorn commented on lib/super_good/solidus_taxjar/discount_calculator.rb:
```diff
@@ -6,7 +6,7 @@ def initialize(line_item)
 
       def discount
-        -line_item.promo_total
+        line_item.adjustments.select { |value| !value.tax? && value.eligible? }.sum(&:amount)
```
```ruby
line_item.adjustments.non_tax.eligible.sum(:amount)
```

Noah-Silvera replied:
Not going to implement this, as discussed in the weekly checkin, as we want this calculator to be able to handle in memory orders.

nvandoorn reviewed (commented):
Small suggestion that came up during today's mob session. This will save us from filtering in Ruby code and use SQL queries to find the appropriate adjustments.

Noah-Silvera commented:
Just a rebase on master!

@forkata just need another review from you as changes were requested!

Noah-Silvera committed:
Use factories instead of directly creating models

This spec directly created some models with active record. We should use
the built in factories instead.

forkata reviewed (approved):
It looks like we need to update some specs here to take into account the new calculation we are doing, but otherwise this looks good to me, thanks @Noah-Silvera!

Noah-Silvera commented:
Drafting this for now

In order to get the `api_params_spec` working, a major refactor to the spec to actually use the calculated values from adjustments and promotions on the order needed to be undertaken.

This work is not done yet - I've pushed a WIP which recalculates promotions correctly but tax is going to be a bigger problem. 

There was also a bug where the sign of the value in the discount calculator was flipped

benjaminwil reviewed (approved):
Beautiful !! :chefs-kiss:

forkata reviewed (approved):
Pending this fixes our issues when re-calculating tax after a `UnitCancel` adjustment is created, this looks good to me 👍🏼 Thanks @Noah-Silvera and Co.

nvandoorn commented:
Gave this a test and it does resolve the issue described in #184
